src/main/java/bio/terra/service/dataset/DatasetService.java

```diff
@@ -353,6 +355,91 @@ public String addDatasetAssetSpecifications(
         .submit();
   }
 
+  public void validateDatasetAssetSpecification(Dataset dataset, AssetModel assetModel) {
+    // Validate Root Table
+    DatasetTable rootTable =
+        dataset.getTables().stream()
+            .filter(datasetTable -> datasetTable.getName().equals(assetModel.getRootTable()))
+            .findFirst()
+            .orElseThrow(
+                () ->
+                    new InvalidAssetException(
+                        "Root table " + assetModel.getRootTable() + " does not exist in dataset."));
+    // Validate Root Column
+    if (!rootTable.getColumns().stream()
+        .map(c -> c.getName())
+        .toList()
+        .contains(assetModel.getRootColumn())) {
+      throw new InvalidAssetException(
+          "Root column "
+              + assetModel.getRootColumn()
+              + " does not exist in table "
+              + rootTable.getName());
+    }
+    // Validate Tables
+    assetModel
+        .getTables()
+        .forEach(
+            assetTable -> {
```

Comment:
As much as I like `forEach`, I think there are places where using it makes the code harder to read. In this case, using ` for` loop is the same amount of code and results in the inner code being indented less, which I think makes it easer to read.
```suggestion
    for (var assetTable : assetModel.getTables()) {
```